gateway-spi/src/main/java/org/apache/knox/gateway/dispatch/DefaultDispatch.java

```diff
@@ -138,6 +144,10 @@ protected HttpResponse executeOutboundRequest( HttpUriRequest outboundRequest )
         }
       }
       auditor.audit( Action.DISPATCH, outboundRequest.getURI().toString(), ResourceType.URI, ActionOutcome.SUCCESS, RES.responseStatus( statusCode ) );
+    } catch( SocketTimeoutException e ) {
+        //Set a 504 instead of throwing an IO exception in the event of a socket timeout,
+        //as an IO exception forces a 500 to be displayed.
```

Comment:
So we need an audit and log line here just like the generic `Exception` catch block. 

The other thing we need to do is make sure this doesn't break the HA case. There a failover needs to happen.

Reply:
Thanks for the feedback. I'll add in the audit and log lines, but just to confirm, when does a socket timeout usually occur? So far all of the socket timeouts that have occurred to me have been caused artificially by lowering the socketTimeout parameter (so I don't have any information on this), but I've been told that if a socket timeout occurs on one node, it is likely to occur on other nodes. Are there situations in which this isn't the case?

Reply:
Yes one server could be dying or overloaded where the second one isn't. That would cause socket timeout on one server but failover would work correctly.

There can definitely be socket timeouts in active/passive and active/active case.

active passive - active server gets overloaded
active/active - could be something wrong on one of the two hosts

Yes you could failover only to have to wait again but that is what failover is designed to do. You can't just skip failover in some cases since you don't have enough information to say whether the failover will succeed or not. The fix for this would be some sort of active list of backend hosts, but Knox doesn't have that today. That is a bigger feature for sure.

Reply:
@risdenk , for active/passive, it doesn't make sense to failover from active to passive in the case of socket timeout. For active/active, it may have value to failover. However, I have 2 arguments against this:
1) Usually the client side that talks to knox has its own timeout (think about a Jupyter notebook), so letting service failover in the case of socket timeout multiple times would eventually cause client side timeout.
2) Most likely the timeout is because the request itself took too much time exceeding the socket timeout config (considering Livy session creation), so failover to another server would also cause a socket timeout as well.

And for active-active type of service, if really needed we can always retry at the client side. 

I think letting client side know that the request caused timeout is very important, instead of Letting them face a generic 500 error without knowing what to do next.

Or we could

Reply:
@shanyu you are arguing for changing the semantics of failover when this change is all about just handling `SocketTimeoutException` as a 504 instead of a 500.

Failover is a completely separate mechanism. If the topology is not configured for failover then Knox won't failover. The `SocketTimeoutException` handling shouldn't be any different in the case of a failover. It needs to tell Knox that a failure happened. Not just assume it knows best and ignore the possibility to failover.

Whether the current failover logic is correct or not is a completely separate issue from handling `SocketTimeoutException`. 

> I think letting client side know that the request caused timeout is very important, instead of Letting them face a generic 500 error without knowing what to do next.

As originally proposed, yes handling `SocketTimeoutException` and setting the response to 504 makes sense, but it must do so in the framework that exists for failover today.